File: backend/src/main/java/com/example/mypixel/model/node/FloorNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.model.Parameter;
import com.example.mypixel.model.ParameterType;
import com.example.mypixel.model.ParamsMap;
import com.example.mypixel.model.Widget;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.lang.NonNull;

import java.util.Map;

@MyPixelNode("Floor")
public class FloorNode extends Node {

    @JsonCreator
    public FloorNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }


    @Override
    public Map<String, Parameter> getInputTypes() {
        return ParamsMap.of(
                "input", Parameter.builder()
                        .type(ParameterType.DOUBLE)
                        .required(true)
                        .widget(Widget.INPUT)
                        .build()
        );
    }
    @Override
    public Map<String, Object> getDefaultInputs() {
        return ParamsMap.of(
                "input", 0.0
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return ParamsMap.of("output", Parameter.required(ParameterType.DOUBLE));
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return ParamsMap.of(
                "category", "Math",
                "description", "Returns the largest integer less than or equal to the input number.",
                "color", "#BA68C8",
                "icon", "FloorIcon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        Map<String, Object> outputs;
        double number = (double) inputs.get("input");
        outputs = Map.of("output", Math.floor(number));
        return outputs;
    }

    @Override
    public void validate() {

    }
}

File: node/nodes/floor_node.py
from typing import Dict, Any
from node import Node
import math


class FloorNode(Node):

    node_type = "Floor"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "input": {
                "type": "DOUBLE",
                "required": True,
                "widget": "INPUT",
                "default": 0.0
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "output": {
                "type": "DOUBLE",
                "required": True,
                "widget": "LABEL"
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "Math",
            "description": "Returns the largest integer less than or equal to the input number.",
            "color": "#BA68C8",
            "icon": "FloorIcon"
        }

    def exec(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        number = inputs.get("input", 0.0)


        try:
            number = float(number)
        except (TypeError, ValueError):
            number = 0.0

        return {"output": math.floor(number)}

    def validate(self, inputs: Dict[str, Any]) -> None:
        pass
